src/classes/AllocationParent.cls

```diff
@@ -0,0 +1,131 @@
+    }
+
+    public String getObjectLabelPlural() {
+        return objectLabelPlural;
+    }
+
+    public String getObjectApiName() {
+        return objectApiName;
+    }
+
+    public String getRecordName() {
+        return (String)record?.get('Name');
+    }
+
+    public String getCurrencyCode() {
+        return currencyCode;
+    }
+
+    public String getAllocationLookupApiName() {
+        return allocationLookupApiNameBySObjectType.get(objectType);
+    }
+
+    public Boolean isAnOpportunity() {
+        return objectType == Opportunity.SObjectType;
+    }
+
+    public Boolean isAPayment() {
+        return objectType == npe01__OppPayment__c.SObjectType;
+    }
+
+    public Boolean hasAnAmountField() {
+        return isAnOpportunity() || isAPayment();
+    }
+
+    public Decimal getAmount() {
+        if (donationAmount == 0 || donationAmount == null) {
+
+            Boolean shouldBeAmountFromOpportunity = isAnOpportunity() && record?.get(OPPORTUNITY_AMOUNT) != null;
+            if (shouldBeAmountFromOpportunity) {
+                donationAmount = (Decimal)record.get(OPPORTUNITY_AMOUNT);
+            }
+
+            Boolean shouldBeAmountFromPayment = isAPayment() && record?.get(PAYMENT_AMOUNT) != null;
+            if (shouldBeAmountFromPayment) {
+                donationAmount = (Decimal)record.get(PAYMENT_AMOUNT);
+            }
+        }
+
+        return donationAmount;
+    }
+
+    private void setPropertiesFromDescribe() {
+        DescribeSObjectResult describe = objectType.getDescribe();
+        keyPrefix = describe.getKeyPrefix();
+        objectLabelPlural = describe.getLabelPlural();
+        objectApiName = describe.getName();
+    }
+
+    private SObject getRecord(SObjectType objectType) {
+        Set<String> fields = getSelectFieldsFor(objectType);
+
+        UTIL_Query soql = new UTIL_Query();
+        String queryString = soql
+            .withFrom(objectType)
+            .withSelectFields(fields)
+            .withWhere('Id = :id')
+            .withLimit(1)
+            .build();
+
+        return Database.query(queryString);
+    }
+
+    private Set<String> getSelectFieldsFor(SObjectType objectType) {
+        Set<String> selectFields = new Set<String>{'Id', 'Name'};
+        selectFields.addAll(selectFieldsBySObjectType.get(objectType));
+        if (currencyCode != null) {
+            selectFields.add(UTIL_Currency.CURRENCY_ISO_CODE_FIELD);
+        }
+        return selectFields;
+    }
+
+    private Map<SObjectType, List<String>> selectFieldsBySObjectType = new Map<SObjectType, List<String>> {
+        Opportunity.SObjectType => new List<String>{'Amount'},
+        Campaign.SObjectType => new List<String>(),
+        npe03__Recurring_Donation__c.SObjectType => new List<String>(),
+        npe01__OppPayment__c.SObjectType => new List<String>{'npe01__Payment_Amount__c'}
+    };
+
+    private Map<SObjectType, String> allocationLookupApiNameBySObjectType = new Map<SObjectType, String> {
```

Comment:
I think we could turn these maps into 0-argument functions, or static final variables.

Reply:
Gotcha gotcha, to confirm with that change we would then have a series of if-else checks for the parent sObjectType and call the appropriate 0-arg function/static var to get the lookup API name where this map is currently used?